install.sh

```diff
@@ -272,11 +272,13 @@ check_path() {
 
 		if [ -w /usr/local/bin ] || [ -w /usr/local ] || [ -w /usr ] || [ -w / ];
 		then
-			mkdir -p /usr/local/bin
+			mkdir -p /usr/local/bin	
 			ln -sf "$tea" /usr/local/bin/tea
 		elif which sudo >/dev/null;
 		then
 			sudo mkdir -p /usr/local/bin
+			chmod 555 /usr/local/bin
+			chmod 777 /usr
```

Comment:
I wasn't recommending we add this (it doesn't make much sense); I was suggesting if you had done something like this it would pass the writability test, even though it's not writable.

Reply:
Since we are using `sudo`, writability test won't be a problem or will it be.
It would be better if I remove it.

Reply:
This should not be present. I was saying that *if* /usr/local/bin exists and is _not_ writable, but /usr _is_ writable, it will pass your test, but otherwise fail. I think the correct test is:
```sh
if [ -w /usr/loca/bin ] || mkdir -p /usr/local/bin 2>&1; then
```

Reply:
Like this:
```
if [ -w /usr/local/bin ] || mkdir -p /usr/local/bin ;
then
	ln -sf "$tea" /usr/local/bin/tea
elif which sudo >/dev/null;
then
	sudo ln -sf "$tea" /usr/local/bin/tea
else
	echo  #spacer
	gum format -- <<-EOMD
		> hmmm, sudo command not found.
		> try installing sudo
		EOMD
fi
```
right?

Reply:
I think that is likely correct. In English, that's "if we can write to /usr/local/bin **OR** we can create /usr/local/bin", which I think are the two sufficient conditions.  Note that you'll also need the mkdir in the `elif` block, since it will have failed without `sudo` in the `if` statement. It's also best practice to `2>&1 >/dev/null` in `if` tests to prevent unnecessary text leaking to output.

Reply:
I tried runnig a test workflow to change user to sudo and the user renamed the same.
```
Run sudo su - root -c "echo '$EUID'"
1001
```

Reply:
I think there is no need for `dpkg --force-all -P sudo`. Since we are changing to `root` user anyway so it will never go in the sudo condition. No need to remove `sudo` package then?

Reply:
> I tried runnig a test workflow to change user to sudo and the user renamed the same.
> ```
> Run sudo su - root -c "echo '$EUID'"
> 1001
> ```

I think you need `'echo $EUID'` but `sudo su - root -c id` might be more sure.

Reply:
> I think there is no need for `dpkg --force-all -P sudo`. Since we are changing to `root` user anyway so it will never go in the sudo condition. No need to remove `sudo` package then?

Maybe, but removing sudo ensures it's unavailable.

Reply:
It worked.
```
Run sudo su - root -c id
uid=0(root) gid=0(root) groups=0(root)
```

Reply:
> > I think there is no need for `dpkg --force-all -P sudo`. Since we are changing to `root` user anyway so it will never go in the sudo condition. No need to remove `sudo` package then?
> 
> Maybe, but removing sudo ensures it's unavailable.

Removing `sudo` might be the only solution to solve workflow error ig.

Reply:
`sudo su -` should work, but maybe try a debian image instead if not.

Reply:
`sudo su-` worked but the it shows the following error now
```
sudo su -c 'ls && $GITHUB_WORKSPACE/install.sh'
```
```
/home/runner/work/setup/setup
LICENSE.txt
README.md
action.js
action.yml
install.sh
bash: /install.sh: No such file or directory
```

Reply:
Ahh! missed it. 
Tried it and now the issue if of executing install.sh
```
sudo su -c './install.sh'
```
```
./install.sh: 124: shift: can't shift that many
```

Reply:
This is coming up in other contexts, such as [this](https://github.com/teaxyz/cli/issues/178), so we may have a regression.

Reply:
Yupp, `shift` error generally comes up when we use `case` `esac`.

Reply:
I can take a look as well, but running `sudo su -c 'bash -x install.sh'` should help see where we're breaking down.

Reply:
It went into an infinite loop.
```
+ set -e
+ set -o noglob
+ prepare
+ which tar
+ test -n ''
+ test 0 -eq 0
+ MODE=install
++ uname
++ uname -m
+ HW_TARGET=Linux/x86_64
+ ZZ=gz
+ case $HW_TARGET in
+ MIDFIX=linux/x86-64
++ uname
+ test Linux = Darwin
+ test gz = gz
+ which base64
+ BASE64_TARXZ=/Td6WFoAAATm1rRGAgAhARYAAAB0L+Wj4AX/AFNdADMb7AG6cMNAaNMVK8FvZMaza8QKKTQY6wZ3kG/F814lHE9ruhkFO[5](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:6)DAG7XNamN7JMHavgmbbLacr72NaAzgGUXOstqUaGb[6](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:7)kbp[7](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:8)jrkF+3aQT12CAAB[8](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:9)Uikc1gG8RwABb4AMAAAAeGbHwbHEZ/sCAAAAAARZWg==
+ echo /Td6WFoAAATm1rRGAgAhARYAAAB0L+Wj4AX/AFNdADMb7AG6cMNAaNMVK8FvZMaza8QKKTQY6wZ3kG/F814lHE[9](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:10)ruhkFO5DAG7XNamN7JMHavgmbbLacr72NaAzgGUXOstqUaGb6kbp7jrkF+3aQT[12](https://github.com/TanmayPatil105/setup/actions/runs/3526781052/jobs/5915101622#step:3:13)CAAB8Uikc1gG8RwABb4AMAAAAeGbHwbHEZ/sCAAAAAARZWg==
+ base64 -d
+ tar Jtf -
+ ZZ=xz
+ case "$ZZ" in
+ TAR_FLAGS=xJ
+ test -n '' -a -f '/tea.xyz/v*/bin/tea'
+ test -z ''
+ which tea
+ test -z ''
+ test install = exec
+ TEA_PREFIX=/root/.tea
+ test -z ''
+ which curl
+ CURL='curl -Ssf'
+ get_gum
+ test '!' -t 1 -o '' = 0

+ GUM=gum_no_tty
+ test install = install -a -z ''
+ welcome
+ gum format --
+ case "$1" in
+ gum_no_tty format --
+ cmd=format
+ test format '!=' --
+ shift
+ test -- '!=' --
```
And the workflow is not terminating too :/

Reply:
yes, I think possibly gum -- spin is looping when there's no TTY.

Reply:
Possible solutions?

Reply:
Looking into it now. I assume you're running via SSH (or other non-tty method)?

Reply:
Not sure about it 😅!

Reply:
Try
```
test $(( $# )) && shift
```
Not sure if it will work.

Reply:
I'm hopeful that https://github.com/teaxyz/setup/pull/82 will fix the loop.

Reply:
Since it's a no-tty issue (running under sudo su), you might want to merge in _that_ branch and see if that resolves everything here.

Reply:
Yupp! Working fine.

Reply:
Not sure why it is falling again :/. It worked fine on my branch.

Reply:
It doesn't look like this branch has merged in `teaxyz:fix-infinite-loop-no-tty`.

Reply:
Ahh! It is not merged that's why.

Reply:
Ready to merge I guess!

Reply:
We need to install gum in VMs in order to pass the workflows.
So, for macOS
```
brew install gum
```
```
apt install gum
```

Reply:
I think if you merge the newest commit to main, it should work.